File: algorithms/0008.string-to-integer-atoi/solution.go
func myAtoi(s string) int {
	fields := strings.Fields(s)
	if len(fields) == 0 {
		return 0
	}

	sign, total, start := 1, 0, 0
	str := fields[0]
	first := string(str[0])
	if first == "+" || first == "-" {
		switch first {
		case "+":
			sign = 1
		case "-":
			sign = -1
		}
		start++
	}

	digits := []string{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"}
	for i := start; i < len(str); i++ {
		s, digit := string(str[i]), 0

		if contains(digits, s) {
			digit, _ = strconv.Atoi(s)
		} else {
			return sign * total
		}

		total = total*10 + digit

		if sign*total > math.MaxInt32 {
			return math.MaxInt32
		} else if sign*total < math.MinInt32 {
			return math.MinInt32
		}
	}

	return sign * total
}

func contains(arr []string, x string) bool {
	for _, v := range arr {
		if v == x {
			return true
		}
	}
	return false
}

File: algorithms/0008.string-to-integer-atoi/solution.rb
# @param {String} s
# @return {Integer}
def my_atoi(s)
  str = s.strip.split(' ').first
  return 0 if str == nil || s.length == 0

  sign, total, start = 1, 0, 0
  # check signs
  if str[0] == '+' || str[0] == '-'
    sign = str[0] == '+' ? 1 : -1
    start += 1
  end

  digits = ('0'..'9').to_a
  (start...str.length).each do |i|
    digit = if digits.include?(str[i])
        Integer(str[i])
      else
        return sign * total
      end

    total = total * 10 + digit
    return 2_147_483_647 if sign * total > 2_147_483_647
    return -2_147_483_648 if sign * total < -2_147_483_648
  end

  sign * total
end
